### PR TITLE
Handle time-outs and other errors when submitting exams

### PR DESCRIPTION
**Pull request by FrankeLukas**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context

To the user it is currently not visible that Artemis is doing work in the background. Additionally we need to make sure it is possible to submit again when a time-out or different error occurs.

### Description

Introduces a client-side time-out (20s) and re-enables submission if that or an error occurs. Adds support for properly displaying 409 errors and introduces a spinner for when the REST call in the background loads. The "Continue" button is also disabled while loading.

### Steps for Testing

Create an Exam and test handing in both in a test run and in a "normal" run. Make sure you can submit again in case the call doesn't reach the server and that the client loads the existing exam if it reached the server but the response didn't make it to the client. Make sure that submitting after time does not work. (Note that test runs can always be submitted after time.)

Test branches that can be deployed:
- Test a delay in client -> server: https://github.com/ls1intum/Artemis/tree/polish/exam/confirm-end-server-timeout
- Test a delay in server -> client: https://github.com/ls1intum/Artemis/tree/polish/exam/confirm-end-client-timeout

### Screenshots

While (up)loading the submission a spinner indicates something is happening:
grafik

An error is displayed for time-outs:
grafik

### CODE REVIEW

File: src/main/webapp/app/exam/participate/exam-participation.service.ts [hunk 1]
```diff
@@ -106,18 +106,23 @@ export class ExamParticipationService {
...
+                    if (error.status === 403 && error.headers.get('x-null-error') === 'error.submissionNotInTime') {
+                        return throwError(new Error('studentExam.submissionNotInTime'));
+                    } else if (error.status === 409 && error.headers.get('x-null-error') === 'error.alreadySubmitted') {
+                        return throwError(new Error('studentExam.alreadySubmitted'));
+                    } else {
+                        return throwError(new Error('studentExam.handInFailed'));
+                    }
+                }),
+            )
+            .delay(30000);
```
Comment (FrankeLukas) [inline src/main/webapp/app/exam/participate/exam-participation.service.ts:125]: This is a dummy testcase that needs to be removed before merging.

**Review by marvinchimeka — APPROVED** [review]
Tested in testing session. Everything works fine. We turned off our internet connection to simulate the situation

**Review by simonzilker — DISMISSED** [review]
Tested in testing session. What  @marvinchimeka said

**Review by philgabr — APPROVED** [review]
Same as @marvinchimeka

**Review by valeriebucher — APPROVED** [review]
Same as @marvinchimeka.

**Review by airelawaleria — CHANGES_REQUESTED** [review]
Tested the client-timeout test brach with @JohannesWeiss 
We've noticed some issues.

After resubmitting the exam after timeout, the user didn't get a summary of the exam (only after the re-loading of the page, the summary was displayed)
image

If the user didn't submit the exam, the summary appears on the screen and disappers after re-loading the page

**Review by JohannesWeiss — COMMENTED** [review]
Tested with @airelawaleria on ts1. Can confirm the mentioned issues.

**Review by derLalla — DISMISSED** [review]
I have only looked through the code and am approving

**Comment by FrankeLukas** [comment]
> After resubmitting the exam after timeout, the user didn't get a summary of the exam

Yes, that expected behaviour as long as it always displays the correct error message. (Since the client will always forcefully run into a timeout on that branch.)

**Review by airelawaleria — APPROVED** [review]
approving since the behavior is expected

**Review by JohannesWeiss — DISMISSED** [review]

**Review by krusche — APPROVED** [review]

**Review by FuchsDominik — DISMISSED** [review]
Code looks good to me :)

**Comment by FrankeLukas** [comment]
Added a comment as requested on Slack.

**Review by krusche — APPROVED** [review]

**Pull request merged**